openedx/core/djangoapps/programs/utils.py

```diff
@@ -461,57 +462,70 @@ def _attach_course_run_upgrade_url(self, run_mode):
     def _attach_course_run_may_certify(self, run_mode):
         run_mode['may_certify'] = self.course_overview.may_certify()
 
-    def _check_enrollment_for_user(self, course_run):
-        applicable_seat_types = self.data['applicable_seat_types']
-
-        (enrollment_mode, active) = CourseEnrollment.enrollment_mode_for_user(
-            self.user,
-            CourseKey.from_string(course_run['key'])
+    def _filter_out_courses_with_entitlements(self, courses):
+        course_uuids = set(course['uuid'] for course in courses)
+        # Filter the entitlements' modes with a case-insensitive match against applicable seat_types
+        query = Q()
+        for seat_type in self.data['applicable_seat_types']:
+            query = query | Q(mode__iexact=seat_type)
+        query = Q(course_uuid__in=course_uuids) & query
+        entitlements = self.user.courseentitlement_set.filter(query)
+        courses_with_entitlements = set(str(entitlement.course_uuid) for entitlement in entitlements)
+        return [course for course in courses if course['uuid'] not in courses_with_entitlements]
+
+    def _filter_out_courses_with_enrollments(self, courses):
+        enrollments = self.user.courseenrollment_set.filter(
+            is_active=True,
+            mode__in=self.data['applicable_seat_types']
         )
+        course_runs_with_enrollments = set(unicode(enrollment.course_id) for enrollment in enrollments)
+        courses_without_enrollments = []
+        for course in courses:
+            if len([run for run in course['course_runs'] if unicode(run['key']) in course_runs_with_enrollments]) == 0:
+                courses_without_enrollments.append(course)
```

Comment:
We should also check for no-id-professional in course runs to make sure v1 with course runs will still work.

Reply:
I made a course run seat with the type no-id-professional, and the response from the programs api had its type and the course run's type listed simply as 'professional', so I don't think it should be an issue with course runs.

Reply:
We should probably normalize `applicable_seat_types` once in `_collect_one_click_purchase_eligibility_data` (remove 'credit', add 'no-id-professional', etc) and then pass it to the filter methods, rather than adding special-case logic everywhere.